core/src/main/java/hudson/util/StreamTaskListener.java

```diff
@@ -166,12 +167,22 @@ public Charset getCharset() {
     private void writeObject(ObjectOutputStream out) throws IOException {
         out.writeObject(new RemoteOutputStream(new CloseProofOutputStream(this.out)));
         out.writeObject(charset==null? null : charset.name());
+        LOGGER.log(Level.FINE, null, new Throwable("serializing here with AUTO_FLUSH=" + AUTO_FLUSH));
     }
 
+    /**
+     * Restores eager remote flushing behavior.
+     * By default, remote tasks are expected to call {@link PrintStream#flush} before exiting.
+     * This flag will ensure that no output is lost from tasks which neglect to do so,
+     * at the expense of heavier Remoting traffic and reduced performance.
+     */
+    private static /* not final */ boolean AUTO_FLUSH = SystemProperties.getBoolean(StreamTaskListener.class.getName() + ".AUTO_FLUSH");
+
     private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
-        out = new PrintStream((OutputStream)in.readObject(),true);
+        out = new PrintStream((OutputStream)in.readObject(), AUTO_FLUSH);
```

Comment:
The meat of the change: switching `auoFlush` on the remote copy from `true` to (by default) `false`.